### PR TITLE
feat(OR-1337): update sdk to support native token

### PR DESCRIPTION
**Pull request by 0x6e616d**
- [x] Update ton-bridge to support withdraw/deposit TON as the native token
- [x] Update eth-bridge to emit the correct `value` in `extraData`
- [x] Update standard-bridge to prevent withdraw/deposit TON via this bridge adapter
- [x] Change `predeploy` dependency from `@eth-optimism/contracts` to `@eth-optimism/core-utils` because `@eth-optimism/contracts` was removed
- [x] Update hardhat tasks
- [x] Change dependencies of artifacts to our new package

### CODE REVIEW

File: packages/core-utils/src/optimism/constants.ts
Comment (nguyenzung) [inline packages/core-utils/src/optimism/constants.ts:34]: What do you think if L1TonAddress is configured when init cross-domain-messenger? I think this address is not contant

Comment (0x6e616d) [reply packages/core-utils/src/optimism/constants.ts:34]: yes, let me update this

**Review by nguyenzung — APPROVED** [review]
LGTM!

**Comment by rlgns98kr** [comment]
I'm sorry for late notice,, I think we can change the naming of TON. (TON -> NativeToken)

**Comment by 0x6e616d** [comment]
> I'm sorry for late notice,, I think we can change the naming of TON. (TON -> NativeToken)

Let me update, thank you 🫶

**Comment by 0x6e616d** [comment]
@rlgns98kr I updated the L1 Native Token address, can you re-check this PR?

**Comment by rlgns98kr** [comment]
great.
I think we still have a lot of TON variable names to change.

We hope to finish OR-1257 by this week. So in my opinion there are two options.

@boohyung @nguyenzung @0x6e616d 
- Merge other PRs and change all variable naming in OR-1257.
- Change the variable and function naming of the current version of OR-1257, merge it into main, and then open the PR again into the main branch.


I prefer the second option because it appears that merging into OR-1257 will increase Brave's workload. 🙇🏼‍♂️

**Comment by 0x6e616d** [comment]
> great. I think we still have a lot of TON variable names to change.
> 
> We hope to finish OR-1257 by this week. So in my opinion there are two options.
> 
> @boohyung @nguyenzung @0x6e616d
> 
> * Merge other PRs and change all variable naming in OR-1257.
> * Change the variable and function naming of the current version of OR-1257, merge it into the main, and then open the PR again into the main branch.
> 
> I prefer the second option because it appears that merging into OR-1257 will increase Brave's workload. 🙇🏼‍♂️

I agree, that OR-1257 is increasing code changes so we merge this to the main, after that, other PRs to OR-1257 will change to the main.
That will help us review the PRs easier

**Comment by nguyenzung** [comment]
> > great. I think we still have a lot of TON variable names to change.
> > We hope to finish OR-1257 by this week. So in my opinion there are two options.
> > @boohyung @nguyenzung @0x6e616d
> > 
> > * Merge other PRs and change all variable naming in OR-1257.
> > * Change the variable and function naming of the current version of OR-1257, merge it into the main, and then open the PR again into the main branch.
> > 
> > I prefer the second option because it appears that merging into OR-1257 will increase Brave's workload. 🙇🏼‍♂️
> 
> I agree, that OR-1257 is increasing code changes so we merge this to the main, after that, other PRs to OR-1257 will change to the main. That will help us review the PRs more easily. So after Brave merge OR-1257 to main, I will change this PR to main and replace TON variables

I think Steven wants to merge this PR into OR-1257. And after OR-1257 is merged, we open another PR to change TON var names. Correct me if I  am wrong? @rlgns98kr

**Comment by rlgns98kr** [comment]
> > > great. I think we still have a lot of TON variable names to change.
> > > We hope to finish OR-1257 by this week. So in my opinion there are two options.
> > > @boohyung @nguyenzung @0x6e616d
> > > 
> > > * Merge other PRs and change all variable naming in OR-1257.
> > > * Change the variable and function naming of the current version of OR-1257, merge it into the main, and then open the PR again into the main branch.
> > > 
> > > I prefer the second option because it appears that merging into OR-1257 will increase Brave's workload. 🙇🏼‍♂️
> > 
> > 
> > I agree, that OR-1257 is increasing code changes so we merge this to the main, after that, other PRs to OR-1257 will change to the main. That will help us review the PRs more easily. So after Brave merge OR-1257 to main, I will change this PR to main and replace TON variables
> 
> I think Steven wants to merge this PR into OR-1257. And after OR-1257 is merged, we open another PR to change TON var names. Correct me if I am wrong? @rlgns98kr


Oh, sorry. The explanation was insufficient.

What I want is to merge OR-1257 first. Afterwards, for other PRs, you can open them again based on the main branch.
To merge OR-1257, you need the following:

- TON -> Native Token (for every function name and variable name and comments)
- Other changes included in https://www.notion.so/onther/First-Audit-commit-b1299e2d6d0b31e7626eb06c43fbad5f36b8a1d6-8becbb08eec44fc2888f61097cbbbe2e

**Comment by boohyung** [comment]
I have marked it to draft.

**Comment by rlgns98kr** [comment]
about naming (ETH -> ETH, TON -> NativeToken)

**Comment by rlgns98kr** [comment]
@0x6e616d 
I changed the base branch to main!

File: packages/tokamak/sdk/tasks/approve-deposit-ton.ts
Comment (nguyenzung) [inline packages/tokamak/sdk/tasks/approve-deposit-ton.ts:105]: I think this task is for specific TON

Comment (0x6e616d) [reply packages/tokamak/sdk/tasks/approve-deposit-ton.ts:105]: @nguyenzung I just renamed it for matching with other tasks, use the term "l2NativeToken" for "TON"

Comment (nguyenzung) [reply packages/tokamak/sdk/tasks/approve-deposit-ton.ts:105]: ApproveAndCall is used for TON, it is not a standard function for other tokens. I think it is still OK of changing the name. Anw, I still found some TON in the code. I think we should update them.

**Comment by nguyenzung** [comment]
Great job, @0x6e616d 
I think there are only some TON need to be updated to NativeToken, after that we can merge

**Comment by 0x6e616d** [comment]
> Great job, @0x6e616d I think there are only some TON need to be updated to NativeToken, after that we can merge

You mean we don't keep use "TON", instead of that we use "L2 Native Token", right? If yes, let me update the code

**Comment by nguyenzung** [comment]
> > > > great. I think we still have a lot of TON variable names to change.
> > > > We hope to finish OR-1257 by this week. So in my opinion there are two options.
> > > > @boohyung @nguyenzung @0x6e616d
> > > > 
> > > > * Merge other PRs and change all variable naming in OR-1257.
> > > > * Change the variable and function naming of the current version of OR-1257, merge it into the main, and then open the PR again into the main branch.
> > > > 
> > > > I prefer the second option because it appears that merging into OR-1257 will increase Brave's workload. 🙇🏼‍♂️
> > > 
> > > 
> > > I agree, that OR-1257 is increasing code changes so we merge this to the main, after that, other PRs to OR-1257 will change to the main. That will help us review the PRs more easily. So after Brave merge OR-1257 to main, I will change this PR to main and replace TON variables
> > 
> > 
> > I think Steven wants to merge this PR into OR-1257. And after OR-1257 is merged, we open another PR to change TON var names. Correct me if I am wrong? @rlgns98kr
> 
> Oh, sorry. The explanation was insufficient.
> 
> What I want is to merge OR-1257 first. Afterwards, for other PRs, you can open them again based on the main branch. To merge OR-1257, you need the following:
> 
> * TON -> Native Token (for every function name and variable name and comments)
> * Other changes included in https://www.notion.so/onther/First-Audit-commit-b1299e2d6d0b31e7626eb06c43fbad5f36b8a1d6-8becbb08eec44fc2888f61097cbbbe2e

@0x6e616d Yep, we are planning changing TON to NativeToken. For a task that related to approve-and-call, I think we should keep "TON", but it is okay if we also change TON -> NativeToken

File: packages/tokamak/sdk/src/adapters/eth-bridge.ts
Comment (rlgns98kr) [inline packages/tokamak/sdk/src/adapters/eth-bridge.ts:4]: -> @tokamak-network/contracts

🙇🏼‍♂️

File: packages/tokamak/sdk/src/adapters/eth-bridge.ts
Comment (rlgns98kr) [inline packages/tokamak/sdk/src/adapters/eth-bridge.ts:47]: I think we have to query NativeToken deposits for this.
l1Token: L1StandardBridge.nativeTokenAddress()
l2Token: predeploys.LegacyERC20NativeToken

File: packages/tokamak/sdk/src/adapters/native-token-bridge.ts
Comment (rlgns98kr) [inline packages/tokamak/sdk/src/adapters/native-token-bridge.ts:49]: Thank you very much!

Then, do we get all of ERC20DepostiInitiated not only NativeToken? 🙇🏼‍♂️

Comment (rlgns98kr) [reply packages/tokamak/sdk/src/adapters/native-token-bridge.ts:49]: I think we can get only for
ERC20DepositInitiated(nativeTokenAddress, Predeploys.LEGACY_ERC20_NATIVE_TOKEN, address, ...)

Comment (rlgns98kr) [reply packages/tokamak/sdk/src/adapters/native-token-bridge.ts:49]: nativeTokenAddress
We can get above value using L1StandardBridge.nativeTokenAddress() 🙇🏼‍♂️

File: packages/tokamak/sdk/src/cross-chain-messenger.ts
Comment (rlgns98kr) [inline packages/tokamak/sdk/src/cross-chain-messenger.ts:148]: I think we change the name!
`nativeTokenAddress` we've used it in contract!

File: packages/tokamak/sdk/src/adapters/standard-bridge.ts
Comment (rlgns98kr) [inline packages/tokamak/sdk/src/adapters/standard-bridge.ts:293]: Will you change the name as Brave said ? 😍🙇🏼‍♂️

File: packages/tokamak/sdk/test/cross-chain-messenger.spec.ts
Comment (rlgns98kr) [inline packages/tokamak/sdk/test/cross-chain-messenger.spec.ts:1657]: I think l1Token is the nativeTokenAddress!

Comment (rlgns98kr) [reply packages/tokamak/sdk/test/cross-chain-messenger.spec.ts:1657]: Oh, let me check it... I'm sorry,,,

**Comment by rlgns98kr** [comment]
I have some my faults in the review.

I think it is great as well!
Thank you very much again!
If you change the naming as Brave said, we can merge this PR! 🙇🏼‍♂️

File: packages/tokamak/sdk/src/adapters/standard-bridge.ts
Comment (rlgns98kr) [inline packages/tokamak/sdk/src/adapters/standard-bridge.ts:303]: We can add the `public` in this function as well. 🙇🏼‍♂️

File: packages/tokamak/sdk/tasks/deposit-erc20.ts
Comment (boohyung) [inline packages/tokamak/sdk/tasks/deposit-erc20.ts:19]: How about changing from Artifact__WETH9 to Artifact__WTON?

**Review by boohyung — COMMENTED** [review]
Thank you very much. Could you double check the running results of hardhat tasks? I found some errors. I'll look for the cause with you.

File: packages/tokamak/sdk/tasks/deposit-erc20.ts
Comment (rlgns98kr) [reply packages/tokamak/sdk/tasks/deposit-erc20.ts:19]: WNativeToken!

**Review by rlgns98kr — APPROVED** [review]
Okay, I think we can merge it.
If we find some errors later, we can fix it then!

Thank you, Nam, Theo, Brave!  😍

**Comment by 0x6e616d** [comment]
> Okay, I think we can merge it. If we find some errors later, we can fix it then!
> 
> Thank you, Nam, Theo, Brave! 😍

I've tested again, and SDK still works well. I agree we can merge this PR to main.

**Pull request merged**
